torch/csrc/jit/python/python_dict.cpp

```diff
@@ -0,0 +1,198 @@
+          "__next__",
+          [](ScriptDictKeyIterator& iter) {
+            auto result = iter.next();
+            return toPyObject(result);
+          })
+      .def("__iter__", [](ScriptDictKeyIterator& iter) { return iter; });
+
+  py::class_<ScriptDictIterator>(m, "ScriptDictIterator")
+      .def(
+          "__next__",
+          [](ScriptDictIterator& iter) {
+            auto result = iter.next();
+            return toPyObject(result);
+          })
+      .def("__iter__", [](ScriptDictIterator& iter) { return iter; });
+
+  py::class_<ScriptDict, std::shared_ptr<ScriptDict>>(m, "ScriptDict")
+      .def(py::init([](py::dict dict) {
+        TypePtr type = nullptr;
+
+        if (dict.size() > 0) {
+          // If the source dictionary is nonempty, try to infer its type.
+          auto inferred_type = tryToInferType(dict);
+
+          if (!inferred_type.success()) {
+            std::stringstream ss;
+            ss << "Unable to infer type of dictionary: "
+               << inferred_type.reason();
+            throw JITException(ss.str());
+          }
+
+          type = inferred_type.type();
+        } else {
+          // If is empty, assume the type is Dict[str, Tensor] as is done in
+          // TorchScript code.
+          type = DictType::create(StringType::get(), TensorType::getInferred());
+        }
+
+        auto data = toIValue(std::move(dict), type);
+        return std::make_shared<ScriptDict>(data);
+      }))
+      .def(py::init([](py::dict dict, const TypePtr& type) {
+        auto data = toIValue(std::move(dict), type);
+        return std::make_shared<ScriptDict>(data);
+      }))
+      .def(
+          "__repr__",
+          [](const std::shared_ptr<ScriptDict>& self) {
+            return toPyObject(self->repr());
+          })
+      .def(
+          "__bool__",
+          [](const std::shared_ptr<ScriptDict>& self) {
+            return toPyObject(self->toBool());
+          })
+      .def(
+          "__len__",
+          [](const std::shared_ptr<ScriptDict>& self) {
+            return toPyObject(self->len());
+          })
+      .def(
+          "__contains__",
+          [](const std::shared_ptr<ScriptDict>& self, py::object key) {
+            try {
+              return toPyObject(self->contains(
+                  toIValue(std::move(key), self->type()->getKeyType())));
+            } catch (const py::cast_error& e) {
+              throw py::key_error();
+            }
+          })
+      .def(
+          "__getitem__",
+          [](const std::shared_ptr<ScriptDict>& self, py::object key) {
+            try {
+              auto value = self->getItem(
+                  toIValue(std::move(key), self->type()->getKeyType()));
+              return toPyObject(value);
+            } catch (const std::out_of_range& e) { // Key doesn't exist.
+              throw py::key_error();
+            } catch (const py::cast_error& e) { // Key isn't the right type.
+              throw py::key_error();
```

Comment:
I just found out `TypeError` exists and that's what is thrown when you try to index a list with something that's not an integer. Perhaps that would be more appropriate here?

```suggestion
              throw py::type_error();
```

Reply:
Yeah, `TypeError` seems more appropriate here. I think your test verifies that same exception type is thrown between eager and scripted dict, maybe we should have a test to cover/verify that?

Also, a minor suggestion, this try-catch-catch struct should be broken down into two try-catches. First one for key object to IValue and `TypeError` should be caught there, second one covers key lookup.

Reply:
> Also, a minor suggestion, this try-catch-catch struct should be broken down into two try-catches. First one for key object to IValue and TypeError should be caught there, second one covers key lookup.

Why is that? Doesn't `try-catch` go through all the catch blocks in order until one of them "works"? Similar to `switch`?

Reply:
Oh, sorry for not being clear, it is a really minor concern. 

It seems that you want TypeError to catch exceptions rising from converting Python object for key to IValue, and you want `out_of_range`/`KeyError` to catch non-existent keys in dictionary. But, let's say there is a `out_of_range` rising from casting Python object key to IValue, then the exception would become a `KeyError` while it should really be `TypeError`.

Reply:
Yeah that's true